source/simulator/solver_schemes.cc

```diff
@@ -0,0 +1,744 @@
+                  ExcIndexRange (component, 0, this->n_components));
+
+          if (component < first_component)
+            return 0;
+          else if (component >= first_component + n_object_components)
+            return 0;
+          else
+            {
+              Vector<double> temp(n_object_components);
+              function_object (p, temp);
+              return temp(component - first_component);
+            }
+        }
+
+        void
+        vector_value (const Point<dim>   &p,
+                      Vector<double>     &values) const
+        {
+          AssertDimension(values.size(), this->n_components);
+
+          // set everything to zero, and then the right components to their correct values
+          values = 0;
+          Vector<double> temp(n_object_components);
+          function_object (p, temp);
+          for (unsigned int i = 0; i < n_object_components; i++)
+            {
+              values(first_component + i) = temp(i);
+            }
+        }
+
+      private:
+        /**
+         * The function object which we call when this class's solution() function is called.
+         **/
+        const std_cxx1x::function<void (const Point<dim> &,Vector<double> &)> function_object;
+
+        /**
+         * The first vector component whose value is to be filled by the given
+         * function.
+         */
+        const unsigned int first_component;
+        /**
+         * The number of vector components whose values are to be filled by the given
+         * function.
+         */
+        const unsigned int n_object_components;
+
+    };
+
+  }
+
+
+
+  template <int dim>
+  double Simulator<dim>::assemble_and_solve_temperature (const bool compute_initial_residual,
+                                                         double *initial_residual)
+  {
+    assemble_advection_system (AdvectionField::temperature());
+
+    if (compute_initial_residual)
+      {
+        Assert(initial_residual != NULL, ExcNotImplemented());
+        *initial_residual = system_rhs.block(introspection.block_indices.temperature).l2_norm();
+      }
+
+    const double current_residual = solve_advection(AdvectionField::temperature());
+    current_linearization_point.block(introspection.block_indices.temperature)
+      = solution.block(introspection.block_indices.temperature);
+
+    if ((initial_residual != NULL) && (*initial_residual > 0))
+      return current_residual / *initial_residual;
+
+    return 0.0;
+  }
+
+  template <int dim>
+  std::vector<double> Simulator<dim>::assemble_and_solve_composition (const bool compute_initial_residual,
+                                                                      std::vector<double> *initial_residual)
+  {
+    std::vector<double> current_residual(introspection.n_compositional_fields,0.0);
+
+    if (compute_initial_residual)
+      Assert(initial_residual->size() == introspection.n_compositional_fields, ExcNotImplemented());
```

Comment:
in the current scheme, you probably also want to test that `initial_residual` is non-`NULL`. But see also my other comment further up.

Reply:
Also, shouldn't it be an `ExcInternalError`? I mean, this shouldn't happen, right?